pip/pip-419.md

```diff
@@ -0,0 +1,180 @@
+                .topic("persistent://public/default/my-topic")
+                .subscriptionName("my-subscription2")
+                .messageListener((consumer, message) -> {
+                    // process message
+                    consumer.acknowledgeAsync(message);
+                })
+                .subscribe();
+    }
+}
+
+```
+
+However, I found that they do not have similar message filtering configurations, which means that a subscription will receive all messages from a topic and then filter them at the business layer if necessary. These meaningless messages may cause a lot of bandwidth waste. **Therefore, I think it is very meaningful to implement message filtering on the pulsar broker side.**
+
+# Motivation
+
+As [Background knowledge](#background-knowledge) mentioned, implement message filtering on the pulsar broker side, enrich the pulsar ecosystem, save bandwidth, and help RocketMQ users switch to Pulsar more smoothly. And, I've noticed two similar PRs https://github.com/apache/pulsar/pull/8544 and https://github.com/apache/pulsar/pull/7629, but neither was approved, so I want to try to get this work implemented.
+
+# Goals
+
+1. **Implement message filtering on the pulsar broker side.**
+
+## In Scope
+
+If this PIP is accepted, it will help users to easily filter messages when consumption. 
+
+# Detailed Design
+
+## Design & Implementation Details
+
+**The general idea is to filter based on the properties of the message at pulsar broker.** 
+
+In the following scenario, the producer adds two properties, `messageType` and `messageSource`, when writing messages. For example:
+
+```java
+ // message1
+ TypedMessageBuilder<byte[]> messageBuilder1 = producer.newMessage();
+ messageBuilder1.property("messageType", "file");
+ messageBuilder1.property("messageSource", "QQ");
+ 
+  // message2
+ TypedMessageBuilder<byte[]> messageBuilder2 = producer.newMessage();
+ messageBuilder2.property("messageType", "audio");
+ messageBuilder2.property("messageSource", "wechat");
+```
+
+If a subscription only wants to consume *audio* messages from w*echat*, how should it be implemented? 
+
+We can implement message filtering logic on pulsar broker at [org.apache.pulsar.broker.service.AbstractBaseDispatcher.filterEntriesForConsumer](https://github.com/apache/pulsar/blob/965ef5c14c93ca896ef4c8f34520066285fcf047/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractBaseDispatcher.java#L108)
+
+The filtering engine can choose the following two solutions. **In general, I prefer the SQL-92 solution, which should have better performance. Groovy may have slightly worse performance due to its dynamic characteristics.**
+
+### Solution 1 SQL-92
+
+Message filtering using [SQL-92](https://en.wikipedia.org/wiki/SQL-92) syntax mentioned in https://github.com/apache/pulsar/pull/8544#issuecomment-1064619792
+
+We can write the following SQL expression to filter messages, the pseudo code is as follow:
```

Comment:
Pulsar JMS reuses ActiveMQ's SQL-92 parser for filtering messages
https://github.com/datastax/pulsar-jms/blob/master/pulsar-jms-filters-common/src/main/java/com/datastax/oss/pulsar/jms/selectors/SelectorSupport.java

Reply:
The project has forked the filtering logic from ActiveMQ in this module: https://github.com/datastax/pulsar-jms/tree/master/activemq-filters